horovod/runner/common/util/safe_shell_exec.py

```diff
@@ -89,32 +102,18 @@ def prepend_context(line, rank, prefix):
         )
 
     def write(text):
-        if index is not None:
+        if index is not None and prefix is not None:
             text = prepend_context(text, index, prefix)
         dst_stream.write(text)
         dst_stream.flush()
 
-    line_buffer = ''
-    while True:
-        text = os.read(src_stream.fileno(), 1000)
-        if text is None:
-            break
-
-        if not isinstance(text, str):
-            text = text.decode('utf-8')
-
-        if not text:
-            break
-
-        for line in re.split('([\r\n])', text):
-            line_buffer += line
-            if line == '\r' or line == '\n':
-                write(line_buffer)
-                line_buffer = ''
+    line = src_stream.readline()
```

Comment:
I remember in a previous version of this script we used `readline()`, but the problem we ran into was that it broke Keras progress bars (which do not print newline characters, so no progress was being shown).  Can you check that this is still the case?  I think the MNIST examples should demonstrate this.

Reply:
That makes sense then that we read chunks and flushed them out as they arrived and only prepended context after a new line. However, a progress bar only makes sense when we are not prepending any context, otherwise it would not render correctly. I think context is only prepended for remote command execution, where a progress bar does not make sense.

I will take care of that use case.